satellite/gracefulexit/endpoint.go

```diff
@@ -339,9 +341,25 @@ func (endpoint *Endpoint) doProcess(stream processStream) (err error) {
 			break
 		}
 
-		request, err := stream.Recv()
-		if err != nil {
-			return eofHandler(err)
+		timedCtx, cancel := context.WithTimeout(ctx, 60*time.Second)
```

Comment:
Why is this 60 seconds? Imagine `endpoint.recvTimeout` is 10 minutes. Then the following code could take the `timedCtx` case after 60 seconds and the `stream.Recv()` could still be executing, leading to race conditions or panics.

I see no reason to use a context here because nothing is using it. Can I suggest something like

```go
done := make(chan struct{})
var request *pb.StorageNodeMessage
var recvErr error
go func() {
    request, recvErr = stream.Recv()
    close(done)
}()

timer := time.NewTimer(endpoint.recvTimeout)
select {
case <-timer.C:
    return ...
case <-done:
}
// now we're sure request and recvErr are set
```

Reply:
okay yeah forgot about the unused timedCtx. I think this makes more sense. the `stream.Recv` is receiving a message from the storage node saying whether it successfully transferred a piece to another node, or failed to transfer a piece to another node.

Reply:
So would that mean that the other timeout with bytes per second doesn't matter? Like we're going to give it a minute here so that puts a hard limit over there as well, right?

Reply:
so I maybe made thing needlessly confusing by putting both timeouts in the same PR, but the timeouts are for different interactions-- the bytes per second timeout limits how long a storage node can take to upload a piece to a different storage node. It doesn't have to do with sending the success message to the satellite.

Reply:
But the satellite will only wait for 1 minute to receive the success message, right? So if the node doesn't complete uploading to a different storage node in a minute, would it fail?

Reply:
ah okay, yes it would fail. I think I should make the satellite timeout longer then, but I'm not sure how much longer.

Reply:
do you have any suggestions for handling sends/receives like this with a maximum timeout? I talked to Egon and we thought having this logic around every send/receive could make the code too complicated (and we'd need it around every send/receive to prevent deadlocking when communicating with nodes)

Reply:
do you think it would be possible in the future to have a timeout for inactive connections in drpc?

Reply:
That depends on what “inactive” means. TCP sockets themselves have keep-alive probes that are transparently handled by the kernel, so in that sense, the sockets will be closed when they are truly inactive. I think we do this for outgoing connections already, and we should make sure we set it for incoming as well. 

The main challenge is that as far as I can see we must either 1. require every rpc to complete in some fixed amount of time, so how long do you pick? or 2. a malicious node can just keep the socket open and alive but inactive (this is true for tcp keep alives as well). drpc doesn’t have the context to know how long an rpc should take: only the application logic around that rpc does.

Reply:
ah okay that makes sense! thanks. @egonelbre ^

Reply:
This makes it sound like we might need a wrapper around stream that adds those timeouts.

Reply:
There's probably a better place to discuss this, but the context is supposed to be used to control the timeouts of all of the operations. I'd also be happy to discuss opportunities for different/better apis once we remove grpc support.